ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -543,6 +582,70 @@ public void onInstanceConfigChange(List<InstanceConfig> configs, NotificationCon
       }
     }
 
+    /**
+     * Triggered whenever the IdealState in current data center has changed (for now, it is usually updated by Helix Bootstrap tool).
+     * @param idealState a list of {@link IdealState} that specifies ideal location of replicas.
+     * @param changeContext the {@link NotificationContext} associated.
+     * @throws InterruptedException
+     */
+    @Override
+    public void onIdealStateChange(List<IdealState> idealState, NotificationContext changeContext)
+        throws InterruptedException {
+      if (!idealStateInitialized.get()) {
+        logger.info("Received initial notification for IdealState change from {}", dcName);
+        idealStateInitialized.set(true);
+      } else {
+        logger.info("IdealState change triggered from {}", dcName);
+      }
+      synchronized (notificationLock) {
+        // rebuild the entire partition-to-resource map in current dc
+        partitionToResourceNameByDc.put(dcName, new ConcurrentHashMap<>());
+        for (IdealState state : idealState) {
+          String resourceName = state.getResourceName();
+          for (String partitionStr : state.getPartitionSet()) {
+            partitionToResourceNameByDc.get(dcName).put(partitionStr, resourceName);
+          }
+        }
+        helixClusterManagerMetrics.idealStateChangeTriggerCount.inc();
+      }
+    }
+
+    /**
+     * Triggered whenever there is a change in the list of live instances.
+     * @param liveInstances the list of all live instances (not a change set) at the time of this call.
+     * @param changeContext the {@link NotificationContext} associated.
+     */
+    @Override
+    public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
+      try {
+        logger.trace("Live instance change triggered from {} with: {}", dcName, liveInstances);
+        updateInstanceLiveness(liveInstances);
+        if (!liveStateInitialized.get()) {
+          logger.info("Received initial notification for live instance change from {}", dcName);
+          liveStateInitialized.set(true);
+        }
+        helixClusterManagerMetrics.liveInstanceChangeTriggerCount.inc();
+      } catch (Throwable t) {
+        errorCount.incrementAndGet();
+        throw t;
+      }
+    }
+
+    /**
+     * Triggered whenever the state of replica in cluster has changed. The snapshot contains up-to-date state of all
+     * resources(replicas) in this data center.
+     * @param routingTableSnapshot a snapshot of routing table for this data center.
+     * @param context additional context associated with this change.
+     */
+    @Override
+    public void onRoutingTableChange(RoutingTableSnapshot routingTableSnapshot, Object context) {
+      logger.info("Routing table change triggered from {}", dcName);
+      synchronized (notificationLock) {
```

Comment:
Yeah, I'm also not sure why this needs extra sync.

Reply:
removed